### DISCUSSION TITLE
Loading assets while loading assets?

### DISCUSSION THREAD

**Discussion by sdfgeoff** [bevyengine/bevy#3940] [Q&A] [answered]
When loading a scene file, it has to invoke other loaders - typically this is done by using a "Standin" component that when loaded, removes itself and is replaced.

But one area I can't quite figure out is how to do this with Materials and Textures without a bit of bother.

Lets say I have a file `redMaterial.mat.ron` that contains:
```
{
    "base_color": [1.0, 0.0, 0.0, 1.0],
    "base_color_texture": "path/to/img.jpg"
}
```
And want to load this material.

I had a look at how the GLTF importer handles this and, urgh, it [reads the bytes into an image manually](https://github.com/bevyengine/bevy/blob/main/crates/bevy_gltf/src/loader.rs#L357).As far as I understand it, this means that if two GLTF's reference the same image it'll be loaded from disk twice and into GPU memory twice. It would be nice to support sharing textures between assets, so another solution is desired. 

All this stems from the fact that (as far as I can tell), there isn't a way to load a different resource from inside an AssetLoader.(eg `load_context.load_other_resource("path")` would be amazing)

I can think of a way around it, but it would require more placeholder components, and means that deep recursion would span lots of frames (eg SceneA requires SceneB requires SceneC requires MeshD requires MaterialE requires ImageF).

So, is there a general way to load assets while loading other assets? Is `load_context.load_other_resource("path")` or something similar possible? (I may have a dig at this sometime)

**Comment by mockersf** [comment, marked as answer]
It's possible, but not very obvious. [here](https://github.com/bevyengine/bevy/discussions/3242#discussioncomment-1886010) is an example.

```rust
// Create an asset path to the asset you want loaded
let asset_path = AssetPath::new_ref(path, None);
// Get an handle to the asset at asset_path
let handle = load_context.get_handle(asset_path.clone());

// Use the handle how you want, load other stuff, ...

// Create your loaded asset. mark it as a dependency
let asset = LoadedAsset::new(CustomAsset { .. }).with_dependency(asset_path);

// Set your complete asset as the default one
load_context.set_default_asset(asset);
```

**Reply by sdfgeoff** [reply]
Epic! This is super useful. 

Look at it go, loading a texture like a boss:
![image](https://user-images.githubusercontent.com/13490050/153930531-06cb1ee7-2273-4dbb-a8a8-f0fc730f5546.png)

Thanks
